### PR TITLE
Type-inferred dictionary

### PR DESCRIPTION
**Pull request by sklam**
~~**WIP. DO NOT MERGE**~~

Support usecase like:

```python
from numba import njit
from numba.typed import Dict

@njit
def foo(k, v):
    d = Dict()
    d[k] = v
    return d
```

TODOs:
- [x] implement type-inferred dict for compiled code
- [x] implement type-inferred dict for interpreted code

### CODE REVIEW

**Comment by ehsantn** [comment]
Can't we support regular `dict()` and `{}` syntax? Is return of boxed typed dict and performance implications afterwards a concern or is this just for simplification of initial implementation?

**Comment by sklam** [comment]
Support for `dict` and `{}` should be fine.  I think we should do it in the next PR.

**Comment by sklam** [comment]
Needs to re-run failed tests once CI is fixed

File: numba/dictobject.py [hunk 1]
```diff
@@ -830,7 +830,18 @@ def impl(d, key, value):
...
+        # Handle the precise case.
+        return impl
+    else:
+        # Handle the imprecise case.
+        d = d.refine(key, value)
+        # Re-bind the key type and value type to match the arguments.
+        keyty, valty = d.key_type, d.value_type
+        # Create the signature that we wanted this impl to have.
+        sig = typing.signature(types.void, d, keyty, valty)
+        return sig, impl
```
Comment (sklam) [inline numba/dictobject.py]: This is a new behavior added to `@overload`.  See docstring at `@overload` for details.

**Comment by sklam** [comment]
After discussing with @esc, I realize I need to add a lot more docs around the redirection in typing and lowering.

File: numba/tests/test_dictobject.py [hunk 2]
```diff
@@ -1261,3 +1260,204 @@ def test_disallow_list(self):
...
+            d[k] = w
+            return d
+
+        k, v, w = 123, 32.1, 321
+        d = foo(k, v, w)
+        self.assertEqual(dict(d), {k: w})
+        self.assertEqual(typeof(d).key_type, typeof(k))
+        self.assertEqual(typeof(d).value_type, typeof(v))
+
+    def test_conflicting_value_type(self):
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_dictobject.py]: Add tests.

File: numba/tests/test_dictobject.py [hunk 3]
```diff
@@ -1261,3 +1260,204 @@ def test_disallow_list(self):
...
+        ct, d, n = foo(True)
+        self.assertEqual(ct, 0)
+        self.assertEqual(n, 1)
+        self.assertEqual(dict(d), {1: 2})
+
+        ct, d, n = foo(False)
+        self.assertEqual(ct, 0)
+        self.assertEqual(dict(d), {})
+        self.assertEqual(n, 0)
+
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_dictobject.py]: Done

File: numba/typed/typeddict.py [hunk 4]
```diff
@@ -219,3 +263,50 @@ def unbox_dicttype(typ, val, c):
...
+    See also: `redirect_type_ctor` in numba/target/bulitins.py
+    """
+    dict_ty = cls.instance_type
+    if not isinstance(dict_ty, types.DictType):
+        msg = "expecting a DictType but got {}".format(dict_ty)
+        return  # reject
+    # Ensure the dictionary is precisely typed.
+    if not dict_ty.is_precise():
+        msg = "expecting a precise DictType but got {}".format(dict_ty)
+        raise errors.LoweringError(msg)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/typed/typeddict.py]: It's intentional.  This definition is redirected from a `lower_builtin`.  If a imprecise DictType is encountered, something is wrong in the lowering definition of `redirect_type_ctor`.

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/typed/typeddict.py]: Oh I see, this seems reasonable then, thanks.

File: numba/typing/templates.py [hunk 5]
```diff
@@ -421,33 +421,51 @@ def generic(self, args, kws):
...
+        # Get the overload implementation for the given types
+        pyfunc = self._overload_func(*args, **kws)
+        if pyfunc is None:
+            # No implementation => fail typing
+            self._impl_cache[cache_key] = None, None
+            return None, None
+        if isinstance(pyfunc, tuple):
+            sig, pyfunc = pyfunc
+            args = sig.args
+            cache_key = None            # don't cache
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/typing/templates.py]: Added some improvement here

File: numba/typed/typeddict.py [hunk 6]
```diff
@@ -108,25 +117,53 @@ def _parse_arg(self, dcttype, meminfo=None):
 
     @property
     def _numba_type_(self):
+        if self._dict_type is None:
+            raise TypeError("invalid operation on untyped dictionary")
         return self._dict_type
 
+    @property
+    def _typed(self):
+        """Returns True if the dictionary is not untyped.
```
Comment (stuartarchibald) [inline numba/typed/typeddict.py]:
```suggestion
        """Returns True if the dictionary is typed.
```

File: numba/typing/templates.py [hunk 7]
```diff
@@ -421,33 +421,88 @@ def generic(self, args, kws):
...
+        return impl, args
+
+    def _build_impl(self, cache_key, args, kws):
+        """Build and cache the implementation.
+
+        Given the postitional (`args`) and keyword arguments (`kws`), obtains
+        the `overload` implementation and wrap it in a Dispatcher object.
+        The expected argument types are returned for use by type-inference.
+        The expected argument types are only different from the given argument
+        types if there is a imprecise type in the given argument types.
```
Comment (stuartarchibald) [inline numba/typing/templates.py]:
```suggestion
        types if there is an imprecise type in the given argument types.
```

File: numba/typing/templates.py [hunk 8]
```diff
@@ -421,33 +421,88 @@ def generic(self, args, kws):
...
+        try:
+            impl, args = self._impl_cache[cache_key]
+        except KeyError:
+            impl, args = self._build_impl(cache_key, args, kws)
+        return impl, args
+
+    def _build_impl(self, cache_key, args, kws):
+        """Build and cache the implementation.
+
+        Given the postitional (`args`) and keyword arguments (`kws`), obtains
```
Comment (stuartarchibald) [inline numba/typing/templates.py]:
```suggestion
        Given the positional (`args`) and keyword arguments (`kws`), obtains
```

File: numba/typing/templates.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/typing/templates.py]: thanks.

**Comment by stuartarchibald** [comment]
Thanks for the fixes, couple of typos to correct then this can be marked for merge (CI pending).

**Comment by seibert** [comment]
Azure seems to have skipped the last commit, but it only consists of docstrings, so it should be fine.  Marking as ready to merge.

**Pull request merged**